calblueprint/airlock
# Logout functionality, Token Invalidation and Redis Integration in Typescript ☀️

This PR implements Logout functionality, token invalidation and integrates redis as the token invalidation storage
Implements #8 in Typescript with  suggested comments from the previous pr.

## PR Description
Provides logout functionality for the Airlock server, provides token invalidation and integrates redis as the `tokenStore` where we store invalid tokens.

## Configuration Before Testing
- Create a Base in Airtable
- Create a Redis endpoint by signing up for a hosted Redis Provider (integration -- i can provide a test endpoint)
- Obtain your Airtable api key
- Create a Table where you will store your users
    - Ensure this table has two fields `username` and `password` 
        - Both fields should be Single Line Text
- Fill out the `server/env.example` file in the root directory with the following environment variables — some have already been pre-populated (keep in mind if you want to test for plain-text passwords you will have to change `DISABLE_HASH_PASSWORD` to `true`)
    - `AIRTABLE_API_KEY`=YOUR_API_KEY
    - `AIRTABLE_BASE_ID`=YOUR_BASE_ID
    - `AIRTABLE_ENDPOINT_URL`=https://api.airtable.com/
    - `AIRTABLE_API_VERSION`=0.1.0
    - `AIRTABLE_USER_AGENT`=Airtable.js/0.7.1
    - `AIRTABLE_USER_TABLE`=NAME_OF_YOUR_USER_TABLE
    - `AIRTABLE_USERNAME_COLUMN_NAME`=NAME_OF_YOUR_USERNAME_COLUMN_NAME
    - `AIRTABLE_USER_TABLE`=NAME_OF_YOUR_USER_TABLE
    - `DISABLE_HASH_PASSWORD`=false
    - `SALT_ROUNDS`=PROVIDE_A_NUMBER
    - `CACHE_PORT=REDIS_PORT`
    - `CACHE_HOST=REDIS_ENDPOINT`
    - `CACHE_PASSWORD=REDIS_DB_PASSWORD`

- Rename `env.example` to `.env`
- Fill out Public Key + Private Key
    - In the the `/server/config` folder
    - Generate a set of public and private keys
    - Name these keys as `pub.pem` and `priv.pem`


## Running The Airlock Server



## How to Test
- Download postman (or any tool that you prefer that will allow you to simulate CRUD operations with the server)

## Testing Logout 
**note** (after posting to this route you will not be able to make queries using the same token)
- Make a `POST` request to `http://localhost:4000/:version/:baseId/__airlock_logout__`
- The `BODY` is empty
- Make the post request — if successful you should receive in return
```
{
    "success": true,
    "message": "User successfully logged out"
}
```

## Testing Register
- Make a `POST` request to `http://localhost:4000/:version/:baseId/__airlock_register__`
    - In the `BODY` provide a `username` and `password` in the form of JSON
```
        {"username": "franco",
        "password": "integrate"}
```
- Make the post request — if successful you should receive in return
 ```
{
        "success": true,
        "token": "Jo__tAafsFuF6PU2YpbGWTolrlAjXAWNrudCaYnMrWm3RRxGz-31n721aKoWz8B7T4ofxHm_ZsxuZUoBGWm9_EeZDF_zrBnOrEij-tMi43DXOXByqskKeM5tR20Y8O8JeRygvdRYyayq5wqf5iOAxNxRINEgdB5HHLpG98B-sHQAHfs4UYPvDvCQT6XXHzEhbthYKht3Wmfg",
        "user": {
            "id": "recIOBFLPhcSwjRHQ",
            "fields": {
                "username": "franco"
            },
            "createdTime": "2019-10-14T02:40:45.000Z"
        }
    }
```



## Testing Login
- Make a `POST` request to `http://localhost:4000/:version/:baseId/__airlock_login__` 
    - In the `BODY` provide the same `username` and `password` you used in testing `/register` in the form of JSON
- Make the post request — if successful you should receive in return
   
 ```{
        "success": true,
        "token": "eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6InJlY2RVOVZEOU5acUIyWllDIiwiZmllbGRzIjp7InVzZXJuYW1lIjoiZnJhbmNvIn0sImNyZWF0ZWRUaW1lIjoiMjAxOS0xMC0xNFQwNDoyOTowNi4wMDBaIiwiaWF0IjoxNTcxMDI3MzYzfQ.bqB1EwGbx8-ud61aj_KDljLtYRQL1TS9o7b2e6_ATIQdNzgtFRt",
        "user": {
            "id": "recdU9VD9NZqB2ZYC",
            "fields": {
                "username": "franco"
            },
            "createdTime": "2019-10-14T04:29:06.000Z"
        }
    }
```

## Testing Airtable queries
- **How to Test (you will notice some of these values come from the .env file)**
    - Open up Postman
    - Create a `GET` request
        - the url should be `http://localhost:4000/v0/BASE_ID/TABLE_NAME?`
        - this is preferably the table you are using for users
    - Configure the `Headers` tab to have the following headers and values
        - `token` = “**This value comes from the contents of the token you received from register/login”**
    - Now you can make a `GET` request
        - If successful you should be able to get all of the contents of the table
    
```
{
        "records": [
            {
                "id": "recdU9VD9NZqB2ZYC",
                "fields": {
                    "username": "franco",
                    "password": "$2b$10$HxkyufAwdukUP90GG/SepC4tOfDl1UZViHm"
                },
                "createdTime": "2019-10-14T04:29:06.000Z"
            },
            {
                "id": "recjpfmEM4FasWzP3",
                "fields": {
                    "username": "arpan",
                    "password": "$2b$10$HxkyufAwdukUP90GG/epC4tOfDl1UZV"
                },
                "createdTime": "2019-10-13T23:45:13.000Z"
            }
        ]
    }
